### DISCUSSION TITLE
Filter Button in App Router with searchParams

### DISCUSSION THREAD

**Discussion by BearCooder** [vercel/next.js#56321] [Help] [answered]
### Summary

Hi, 

I have some data in my config file that I map through and return in my page.tsx By default the page shows all images. I have 3 buttons to filter through the images. I take the category from my data config and match it with the button value. When I click a button e.g. Animals then it only should shows images from the animal category and so on. When I click on the button a second time it resets to the default state.

I implemented the filter with searhParams and useRouter to have the state persisted in the URL. The recent video from @leerob on Twitter was very helpful, thanks.

But currently when I click on the button the url changes, however I get a 404 page not found error. I am not sure where my error is. Here in this [Codesandbox](https://codesandbox.io/p/github/BearCooder/FilterButton-AppRouter/main?file=%2Fapp%2Ffilter2.tsx%3A1%2C1&workspaceId=f4b731be-a853-47e0-bbd3-c6f7c81934e4) you can see my code, first I see the content and the buttons. When I click on the button then instead of showing only content from the specific category I get a 404.

Also posting here, the filterbutton code:
```

import { useRouter } from "next/navigation";
import { useSearchParams } from "next/navigation";
import { useCallback } from "react";
import { createUrl } from "./utils";

interface FilterButtonsProps {
  categories: string[];
}

export default function FilterButtons({ categories }: FilterButtonsProps) {
  const router = useRouter();
  const searchParams = useSearchParams();
  const newParams = new URLSearchParams(searchParams.toString());

  const handleFilter = useCallback(
    (category: string) => {
      const currentCategory = newParams.get("category");
      if (currentCategory === category) {
        // Deselect the category
        newParams.delete("category");
      } else {
        // Select the category
        newParams.set("category", category);
      }

      router.push(createUrl("/filter", newParams));
    },
    [router, newParams]
  );

  return (
    <div>
      {categories.map((category) => (
        <button
          key={category}
          onClick={() => handleFilter(category)}
          className={`mr-2 ${
            searchParams.get("category") === category
              ? "bg-blue-500 text-white"
              : "bg-gray-300"
          }`}
        >
          {category}
        </button>
      ))}
    </div>
  );
}
```

Can anyone help me why the filter is not working after clicking the button please?

### Additional information

```js
Also here I created a repository with the code: https://github.com/BearCooder/FilterButton-AppRouter
```

**Comment by icyJoseph** [comment, marked as answer]
Mm but you are pushing to `/filter?category=Architecture` which means that you should have a `filter/page.tsx` in your routing directory.

I see you only have `app/filter.tsx` though, that won't do it. Take a look here: https://stackblitz.com/edit/github-xdjpjn?file=app%2Ffilter%2Fpage.tsx,app%2Fpage.tsx,app%2Ffilter2.tsx

**Reply by BearCooder** [reply]
Oh my god what a stupid mistake! In my case I would just need to use `const target = createUrl('/', newParams);` instead of `const target = createUrl('/filter', newParams);`
I know the routing convention of the app directory but I just need to get used to it I suppose.
Thank you very much!

Another quick question: Is my approach and code for filtering buttons fine or can you recommend some improvements?

**Reply by icyJoseph** [reply]
Nice! 

I guess it is fine, because users can share URLs too and start easily with a given filter. The URL search params are sort of an external store to your app. It is a valid design choice.

Please don't forget to mark as answered :)

**Reply by BearCooder** [reply]
Thank you for the feedback and done! :)
